nn_adapt/solving_one2n.py

```diff
@@ -0,0 +1,244 @@
+"""
+Time dependent goal-oriented error estimation
+"""
+"""
+Functions for solving problems defined by configuration
+files and performing goal-oriented error estimation.
+"""
+from firedrake import *
+from firedrake.petsc import PETSc
+from firedrake.mg.embedded import TransferManager
+from firedrake_adjoint import *
+from firedrake.adjoint import get_solve_blocks
+from pyroteus.error_estimation import get_dwr_indicator
+import abc
+from time import perf_counter
+
+import matplotlib.pyplot as plt
+
+
+tm = TransferManager()
+
+
+class Solver(abc.ABC):
+    """
+    Base class that defines the API for solver objects.
+    """
+
+    @abc.abstractmethod
+    def __init__(self, mesh, ic, **kwargs):
+        """
+        Setup the solver.
+
+        :arg mesh: the mesh to define the solver on
+        :arg ic: the initial condition
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def function_space(self):
+        """
+        The function space that the PDE is solved in.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def form(self):
+        """
+        Return the weak form.
+        """
+        pass
+
+    @abc.abstractmethod
+    def iterate(self, **kwargs):
+        """
+        Solve the PDE.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def solution(self):
+        """
+        Return the solution field.
+        """
+        pass
+
+
+def get_solutions_one2n(
+    mesh,
+    config,
+    solve_adjoint=True,
+    refined_mesh=None,
+    init=None,
+    convergence_checker=None,
+    **kwargs,
+):
+    """
+    Solve forward and adjoint equations on a
+    given mesh.
+
+    This works only for steady-state problems.
+    Trying to work it out.
+
+    :arg mesh: input mesh
+    :arg config: configuration file, which
+        specifies the PDE and QoI
+    :kwarg solve_adjoint: should we solve the
+        adjoint problem?
+    :kwarg refined_mesh: refined mesh to compute
+        enriched adjoint solution on
+    :kwarg init: custom initial condition function
+    :kwarg convergence_checker: :class:`ConvergenceTracer`
+        instance
+    :return: forward solution, adjoint solution
+        and enriched adjoint solution (if requested)
+    """
+    
+    tt_steps = config.parameters.tt_steps
+
+    # Solve forward problem in base space
+    V = config.get_function_space(mesh)
+    out = {"times": {"forward": -perf_counter()}}
+    with PETSc.Log.Event("Forward solve"):
+        if init is None:
+            ic = config.get_initial_condition(V)
+        else:
+            ic = init(V)
+        solver_obj = config.Solver_one2n(mesh, ic, **kwargs)
+        solver_obj.iterate()
+        q = solver_obj.solution
+        # Calculate QoI
+        qoi = 0
+        for step in range(tt_steps):
+            J = config.get_qoi(V)(q[step])
+            qoi += assemble(J)
+        qoi = qoi / tt_steps
```

Comment:
Good! Usually I would compute the QoI during the solver loop, but I think this is okay, too.